src/MessageList/MessageListItem.vala

```diff
@@ -388,13 +388,45 @@ public class Mail.MessageListItem : Gtk.ListBoxRow {
                 return;
             }
 
-            message_content = (string) os.steal_data ();
+            // Convert the message to UTF-8 to ensure we have a valid GLib string.
+            message_content = convert_to_utf8(os, field.param ("charset"));
+
             if (field.subtype == "html") {
                 message_is_html = true;
             }
         }
     }
 
+    private static string convert_to_utf8(GLib.MemoryOutputStream os, string? encoding) {
+        var num_bytes = (int) os.get_data_size ();
+        var bytes = (string) os.steal_data ();
+
+        string? utf8 = null;
+
+        if (encoding != null) {
+            string? iconv_encoding = Camel.iconv_charset_name (encoding);
+            if (iconv_encoding != null) {
+                try {
+                    utf8 = GLib.convert (bytes, num_bytes, "UTF-8", iconv_encoding);
+                } catch (ConvertError e) {
+                    // Nothing to do - result will be assigned below.
+                }
+            }
+        }
+
+        if (utf8 == null || !utf8.validate()) {
+            // If message_content is not valid UTF-8 at this point, assume that
+            // it is ISO-8859-1 encoded by default, and convert it to UTF-8.
+            try {
+                utf8 = GLib.convert (bytes, num_bytes, "UTF-8", "ISO-8859-1");
+            } catch (ConvertError e) {
+                critical("Every string should be valid ISO-8859-1");
```

Comment:
Missing a space between function name and parenthesis opening
You can also include the e.message text in the critical message so that we have clear explanation of why the conversion failed.
You also probably want to return an UTF-8 encoded HTML page containing an `unable to load message` error message and the explanation :man_shrugging:

Reply:
That's a good idea about the error page, but I think I'd prefer to postpone this for now. It would require localization work, and I *think* that this can't happen anyway.